code/modules/religion/sparring/sparring_contract.dm

```diff
@@ -0,0 +1,121 @@
+	data["left_sign"] = left_partner ? left_partner.real_name : "none"
+	data["right_sign"] = right_partner ? right_partner.real_name : "none"
+	return data
+
+/obj/item/sparring_contract/proc/get_possible_areas()
+	var/list/area_names = list()
+	var/datum/religion_sect/spar/sect = GLOB.religious_sect
+	for(var/key in sect.arenas)
+		area_names += key
+	return area_names
+
+/obj/item/sparring_contract/ui_act(action, list/params)
+	. = ..()
+	if(.)
+		return
+
+	var/mob/user = usr
+
+	if(!ishuman(user))
+		to_chat(user, span_warning("This contract refuses to be signed by a lesser creature such as yourself."))
+		return
+
+	var/datum/religion_sect/spar/sect = GLOB.religious_sect
+
+	if(user in sect.past_opponents && params["stakes"] == STANDARD_STAKES)
+		to_chat(user, span_warning("This contract refuses to be signed up for a holy match by a previous holy match loser. Pick a different stake!"))
+
+	//any updating of the terms should update the UI to display new terms
+	. = TRUE
+
+	switch(action)
+		if("fight")
+			var/mob/living/carbon/human/left_partner = signed_by[1]
+			var/mob/living/carbon/human/right_partner = signed_by[2]
+			if(!left_partner || !right_partner || !left_partner.mind || !right_partner.mind)
+				return
+			var/chaplain = left_partner.mind.holy_role ? left_partner : right_partner
+			var/opponent = left_partner.mind.holy_role ? right_partner : left_partner
+			new /datum/sparring_match(weapons_condition, arena_condition, stakes_condition, chaplain, opponent)
+			qdel(src)
+		if("sign")
+			if(user in signed_by)
+				to_chat(user, span_warning("You've already signed one side of the contract."))
+				return
+			var/area/arena_condition_name = GLOB.areas_by_type[arena_condition]
+			arena_condition_name = format_text(arena_condition_name.name)
+			//setting/checking for terms changed
+			var/terms_changed = FALSE
+			if(params["weapon"] != weapons_condition)
+				if(!params["weapon"])
+					return //they hit f5 to clear data then submitted
+				terms_changed = TRUE
+				weapons_condition = params["weapon"]
+			if(params["area"] != arena_condition_name)
+				if(!params["area"])
+					return //they hit f5 to clear data then submitted
+				terms_changed = TRUE
+				var/new_area_condition = sect.arenas[params["area"]]
+				arena_condition = new_area_condition
+			if(params["stakes"] != stakes_condition)
+				if(!params["stakes"])
+					return //they hit f5 to clear data then submitted
+				terms_changed = TRUE
+				weapons_condition = params["stakes"]
+			//if you change the terms you have to get the other person to sign again.
+			if(terms_changed && (signed_by[1] || signed_by[2]))
+				signed_by = list(null, null)
+				to_chat(user, span_warning("You will need to get your sparring partner to sign again under these new terms you've set."))
+			//fluff and signing
+			if(params["sign_position"] == LEFT_FIELD)
+				signed_by[1] = user
+			else
+				signed_by[2] = user
```

Comment:
This will hang references to the people signing the contracts. Should be a weakref

Reply:
See/Read https://github.com/tgstation/tgstation/blob/master/.github/HARDDEL_GUIDE.md